src/librustc_typeck/check/pat.rs

```diff
@@ -81,6 +81,12 @@ struct TopInfo<'tcx> {
     parent_pat: Option<&'tcx Pat<'tcx>>,
 }
 
+/// Returned by `fn check_array_pat_len`.
+enum PatLenOrArray<'tcx> {
+    Pat(u64),
+    Array(Ty<'tcx>),
+}
```

Comment:
This seems unnecessary. You can unconditionally return `Ty<'tcx>` in `check_array_pat_len` instead...